### ISSUE TITLE
hardlinked symlinks are not supported

### ISSUE THREAD

**Issue by ThomasWaldmann** [borgbackup/borg#2379]
borg supports hardlinks and it supports symlinks and both are backupped as it should be.

A problem are "hardlinked symlinks".

```
# on macOS 12
% touch target
% ln -s target symlink1
% ln -P symlink1 symlink2
% ls -li             
40490316 lrwxr-xr-x  2 tw  staff  6 May 11 12:32 symlink1 -> target
40490316 lrwxr-xr-x  2 tw  staff  6 May 11 12:32 symlink2 -> target
40490309 -rw-r--r--  1 tw  staff  0 May 11 12:31 target
```

This part of #2324 is not addressed yet (except that symlinks are not in the documented list of supported hardlinked fs objects).

The problem here is the dual-use of item.source (see #2343), which should be solved first.

Then, hardlinked symlinks could be supported in same way as other hardlinked fs objects.

Note:
- This is a minor issue, likely nobody uses hardlinked symlinks anyway.
- If somebody uses them, the only negative effect is that such hardlinks are not archived as hardlinks (and thus: not restored as hardlinks).

**Comment by enkore** [comment]
- Since permissions on symlinks are ignored and symlinks cannot be changed after creation, the usual benefits of hardlinks don't apply. The only observable effect is st_nlink (and one less inode used).

**Comment by nilsmeyer** [comment]
It's just called a hard link (as opposed to a symbolic link, a link is either hard or symbolic). Only major use case I have seen for it is rsnapshot, which being a backup tool is likely not being backed up with borg.

**Comment by BarbzYHOOL** [comment]
so when saving hardlinks with borg, they are instead soft symlinks? do I understand right?

**Comment by ThomasWaldmann** [comment]
@BarbzYHOOL read the issue title again.

borg supports hardlinks and it supports symlinks and both are backupped as it should be.

The only problem are "hardlinked symlinks".

**Comment by BarbzYHOOL** [comment]
Lol, my bad! hardlinked symlinks, I thought it meant "hardlinks". Also seems like nilsmeyer didn't get it either "It's just called a hard link (as opposed to a symbolic link, a link is either hard or symbolic)."

Now I understand better why you said "almost nobody use them anyway" (but to be fair, most people also say "do'nt use hardlinks ever, use symlinks" so that's why I confused)

What the hell is a hardlinked symlink? a hardlink on a symlink?

**Comment by ThomasWaldmann** [comment]
yes.

**Comment by ThomasWaldmann** [comment]
Working on this in #6663.

Since that uses `item.hlid` to relate hardlinks, there is no issue with usage of `item.source` any more.

**Comment by ThomasWaldmann** [comment]
borg2 branch supports hardlinked symlinks.

**Comment by prahal** [comment]
Is this fixed in borg 1.2? Maybe the FAQ should be updated to stop advertising hardlinked symlinks are not supported?

> For hardlinked symlinks, the hardlinking can not be archived (and thus, the hardlinking will not be done at extraction time). The symlinks will be archived and extracted as non-hardlinked symlinks, see [#2379](https://github.com/borgbackup/borg/issues/2379).

**Comment by ThomasWaldmann** [comment]
This is rather a design limitation of the borg item metadata (in borg < 2) than a bug. It's fixed in master branch, which will be released as borg 2 at some time in the future.
